--- a/src/intel/common/meson.build
+++ b/src/intel/common/meson.build
@@ -1,7 +1,9 @@
 # Copyright © 2017 Intel Corporation
 # SPDX-License-Identifier: MIT
 
-files_libintel_common = files(
+files_libintel_common = []
+if host_machine.system() != 'cubit'
+files_libintel_common += files(
   'i915/intel_engine.c',
   'i915/intel_engine.h',
   'i915/intel_gem.c',
@@ -14,10 +16,16 @@
   'xe/intel_gem.h',
   'xe/intel_queue.c',
   'xe/intel_queue.h',
-  'intel_aux_map.c',
-  'intel_aux_map.h',
   'intel_bind_timeline.c',
   'intel_bind_timeline.h',
+  'intel_gem.c',
+  'intel_gem.h',
+)
+endif
+
+files_libintel_common += files(
+  'intel_aux_map.c',
+  'intel_aux_map.h',
   'intel_buffer_alloc.h',
   'intel_common.c',
   'intel_common.h',
@@ -27,8 +35,6 @@
   'intel_debug_identifier.c',
   'intel_engine.c',
   'intel_engine.h',
-  'intel_gem.c',
-  'intel_gem.h',
   'intel_genX_state_brw.h',
   'intel_genX_state_elk.h',
   'intel_guardband.h',
