--- a/src/intel/common/intel_engine.c
+++ b/src/intel/common/intel_engine.c
@@ -26,6 +26,7 @@
 #include "util/macros.h"
 
 #include "intel_engine.h"
+#if !defined(__cubit__)
 #include "i915/intel_engine.h"
 #include "xe/intel_engine.h"
 
@@ -43,6 +44,8 @@
    }
 }
 
+#endif
+
 int
 intel_engines_count(const struct intel_query_engine_info *info,
                     enum intel_engine_class engine_class)
@@ -76,6 +79,7 @@
    }
 }
 
+#if !defined(__cubit__)
 static bool
 is_guc_semaphore_functional(int fd, const struct intel_device_info *info)
 {
@@ -108,3 +112,4 @@
 
    return supported ? intel_engines_count(engine_info, engine_class) : 0;
 }
+#endif
