--- a/std/src/sys/alloc/mod.rs
+++ b/std/src/sys/alloc/mod.rs
@@ -71,2 +71,5 @@
 cfg_select! {
+    target_os = "none" => {
+        mod cubit;
+    }
     any(
--- a/std/src/sys/io/error/mod.rs
+++ b/std/src/sys/io/error/mod.rs
@@ -1,2 +1,6 @@
 cfg_select! {
+    target_os = "none" => {
+        mod cubit;
+        pub use cubit::*;
+    }
     target_os = "hermit" => {
--- a/std/src/sys/random/mod.rs
+++ b/std/src/sys/random/mod.rs
@@ -1,2 +1,6 @@
 cfg_select! {
+    target_os = "none" => {
+        mod cubit;
+        pub use cubit::fill_bytes;
+    }
     // Tier 1
--- a/std/src/sys/thread_local/mod.rs
+++ b/std/src/sys/thread_local/mod.rs
@@ -24,4 +24,5 @@
 cfg_select! {
     any(
+        target_os = "none",
         all(target_family = "wasm", not(target_feature = "atomics")),
         target_os = "uefi",
@@ -105,3 +106,4 @@
         any(
+            target_os = "none",
             all(target_family = "wasm", not(
                 all(target_os = "wasi", target_env = "p1", target_feature = "atomics")
--- a/std/src/sys/time/mod.rs
+++ b/std/src/sys/time/mod.rs
@@ -1,2 +1,6 @@
 cfg_select! {
+    target_os = "none" => {
+        mod cubit;
+        use cubit as imp;
+    }
     target_os = "hermit" => {
--- a/std/src/sys/stdio/mod.rs
+++ b/std/src/sys/stdio/mod.rs
@@ -3,2 +3,6 @@
 cfg_select! {
+    target_os = "none" => {
+        mod cubit;
+        pub use cubit::*;
+    }
     any(target_family = "unix", target_os = "hermit", target_os = "wasi") => {
